### PR TITLE
Fix activation checkpointing being a no-op during v3 finetuning 

### PR DESCRIPTION
**Pull request by anuragg1209**
## Motivation and Context

Activation checkpointing was effectively a no-op during fine-tuning of v3 (and v2.5/v2.6), causing CUDA OOM on long-context training (e.g. 133k×600 multiclass
   on 4×A100). This PR wires `PerformanceOptions.force_recompute_layer` end-to-end from the fine-tuning trainer to the model so that gradient checkpointing actually fires during the backward pass.                                                                                                                  

The fine-tuning forward path (_training_forward → DDP wrapper → TabPFNClassifier.forward → InferenceEngineBatchedNoPreprocessing.iter_outputs → model(...)) never threaded performance_options through, so the model used its default (force_recompute_layer=False) and stored all activations for backprop → OOM.

## Public API Changes

-   [ ] No Public API changes
-   [ ] Yes, Public API changes (Details below)

---

## How Has This Been Tested?

 Successfully executed end-to-end fine-tuning run with v3, and all pytest test cases passed. 
---

## Checklist

-   [x] The changes have been tested locally.
-   [x] Documentation has been updated (if the public API or usage changes).
-   [x] A changelog entry has been added (see `changelog/README.md`), or "no changelog needed" label requested.
-   [x] The code follows the project's style guidelines.
-   [x] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

File: changelog/917.fixed.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix activation checkpointing being a no-op during finetuning of TabPFN v3 (and v2.5/v2.6), which caused CUDA OOM on long-context training. The legacy `model.recompute_layer` attribute only applied to the v2-base architecture; v3/v2.5/v2.6 honor `PerformanceOptions.force_recompute_layer`, which now flows from `FinetunedTabPFNClassifier`/`FinetunedTabPFNRegressor` (controlled by `use_activation_checkpointing`, default `True`) through `TabPFNClassifier.forward`/`TabPFNRegressor.forward` and `InferenceEngineBatchedNoPreprocessing.iter_outputs` to the model.
```
Comment (bejaeger) [inline changelog/917.fixed.md]:
```suggestion
Adapt `recompute_layer` flag in `FinetunedTabPFNClassifier`/`FinetunedTabPFNRegressor` to new `PerformanceOptions` interface.
```

File: src/tabpfn/classifier.py [hunk 2]
```diff
@@ -1454,10 +1461,14 @@ def forward(  # noqa: C901, PLR0912
             self.executor_.use_torch_inference_mode(use_inference=actual_inference_mode)
 
         outputs = []
+        iter_kwargs: dict[str, Any] = {}
+        if is_batched_for_grads and performance_options is not None:
+            iter_kwargs["performance_options"] = performance_options
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/classifier.py]: We only want to update the `InferenceEngineBatchedNoPreprocessing` as it's the only one used by finetuning functionality and not change everywhere.

File: src/tabpfn/regressor.py [hunk 3]
```diff
@@ -1098,8 +1103,11 @@ def _iter_forward_executor(
             # only these two modes support this option
             self.executor_.use_torch_inference_mode(use_inference=use_inference_mode)
         std_borders = self.znorm_space_bardist_.borders.cpu().numpy()
+        iter_kwargs: dict[str, Any] = {}
+        if is_batched_for_grads and performance_options is not None:
+            iter_kwargs["performance_options"] = performance_options
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/regressor.py]: We only want to update the `InferenceEngineBatchedNoPreprocessing` as it's the only one used by the finetuning functionality and not change everywhere.

File: src/tabpfn/finetuning/finetuned_base.py [hunk 4]
```diff
@@ -340,7 +341,17 @@ def _build_eval_config(
...
+
+        Injects ``performance_options`` derived from
+        ``self.use_activation_checkpointing`` so v3 (and v2.5/v2.6) honor
+        gradient checkpointing during fine-tuning. Architectures silently ignore
+        unsupported fields, so this is safe across versions.
+        """
+        if self.use_activation_checkpointing and "performance_options" not in kwargs:
+            kwargs["performance_options"] = PerformanceOptions(
+                force_recompute_layer=True,
+            )
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py]: I looked into this, on v3, `force_recompute_layer` and `use_chunkwise_inference` are deliberately mutually exclusive: `tabpfn_v3.py:2229` does `force_recompute_layer=force_recompute_layer and is_full_path`, where `is_full_path = not use_chunks` (line 2083). Preserving v3's default `use_chunkwise_inference=True` would set `is_full_path=False` and silently disable the activation checkpointing we're trying to enable. Constructing a fresh `PerformanceOptions(force_recompute_layer=True)` is intentional here — it routes v3 onto the full path so checkpointing actually fires during the backward pass. The other defaults (save_peak_memory_factor=None, enable_torch_compile=False) are also what we want for fine-tuning.

File: src/tabpfn/classifier.py [hunk 5]
```diff
@@ -1403,6 +1408,8 @@ def forward(  # noqa: C901, PLR0912
                 Otherwise, probabilities are returned.
             return_raw_logits: If True, returns the raw logits, without
                 averaging estimators or temperature scaling.
+            performance_options: Optional runtime performance overrides for
+                inference execution.
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/classifier.py]: Updated the docstring, which now explains why it is only passed to the Batched inference engine path.

File: src/tabpfn/regressor.py [hunk 6]
```diff
@@ -1169,6 +1178,8 @@ def forward(
             use_inference_mode: Flag for inference mode., default at False since
             it is called within predict. During FineTuning forward() is called
             directly by user, so default should be False here.
+            performance_options: Optional runtime performance overrides for
+                inference execution.
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/regressor.py]: Addressed in the docstring.

File: src/tabpfn/inference.py [hunk 7]
```diff
@@ -547,6 +547,7 @@ def iter_outputs(
         *,
         autocast: bool,
         task_type: str,
+        performance_options: PerformanceOptions | None = None,
     ) -> Iterator[tuple[torch.Tensor | dict, list[EnsembleConfig]]]:
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/inference.py:554]: Addressed in the above comments. We only want to change the functions which are used by finetuning functionality and not everywhere.

File: src/tabpfn/base.py [hunk 8]
```diff
@@ -366,6 +368,11 @@ def create_inference_engine(  # noqa: PLR0913
             autocast=use_autocast_,
         )
     if fit_mode == "batched":
+        # Finetuning constructs InferenceEngineBatchedNoPreprocessing directly via
+        # fit_from_preprocessed with explicit PerformanceOptions; this branch is
+        # only reachable from the standard fit() path (currently guarded by a
+        # fit_mode auto-switch in classifier/regressor), so default options are
+        # appropriate here.
```
Comment (bejaeger) [inline src/tabpfn/base.py]: Let's raise a ValueError instead and say that `InferenceEngineBatchedNoPreprocessing` should be initialized directly without using this factory.

Gets us closer to disentangling the FT code.

File: src/tabpfn/regressor.py [hunk 9]
```diff
@@ -1190,7 +1198,10 @@ def forward(
         borders: list[np.ndarray] = []
 
         for border, output in tqdm(
-            self._iter_forward_executor(X, use_inference_mode=use_inference_mode),
+            self._iter_forward_executor(
+                X,
+                use_inference_mode=use_inference_mode,
+            ),
```
Comment (bejaeger) [inline src/tabpfn/regressor.py]: can be reverted now to a one-liner :)

**Review by bejaeger — APPROVED** [review]
Nice! Just two more things and we're good to go!

thanks for fixing.

File: src/tabpfn/regressor.py (hunk 9 above)
Comment (anuragg1209) [reply src/tabpfn/regressor.py]: Ruff always acts against me. Will fight it!
